File: solutions/haskell/state-of-tic-tac-toe/2/src/StateOfTicTacToe.hs
module StateOfTicTacToe (gameState, GameState (..)) where

data GameState = WinX | WinO | Draw | Ongoing | Impossible deriving (Eq, Show)

gameState :: [String] -> GameState
gameState board
  | countX > 1 && countO == 0
      || countO > countX
      || win 'X' && (countX <= countO)
      || win 'O' && (countO < countX) =
      Impossible
  | win 'X' = WinX
  | win 'O' = WinO
  | ' ' `notElem` flat = Draw
  | otherwise = Ongoing
 where
  flat = concat board
  countX = length . filter (== 'X') $ flat
  countO = length . filter (== 'O') $ flat
  win p =
    or
      [ (p, p, p) == (flat !! a, flat !! b, flat !! c)
      | (a, b, c) <-
          [ (0, 1, 2) -- lines
          , (3, 4, 5)
          , (6, 7, 8)
          , (0, 3, 6) -- cols
          , (1, 4, 7)
          , (2, 5, 8)
          , (0, 4, 8) -- diags
          , (2, 4, 6)
          ]
      ]

File: solutions/lua/state-of-tic-tac-toe/2/state-of-tic-tac-toe.lua
local function win(s, v)
	local positions = {
		("^%s%s%s......"):format(v, v, v),
		("^...%s%s%s..."):format(v, v, v),
		("^......%s%s%s"):format(v, v, v),
		("^%s...%s...%s"):format(v, v, v),
		("^..%s.%s.%s.."):format(v, v, v),
		("%s..%s..%s"):format(v, v, v),
	}
	for _, p in ipairs(positions) do -- any match
		if s:match(p) then return true end
	end
end
return {
	gamestate = function(board)
		local flat = table.concat(board)
		local countX = select(2, flat:gsub("X", ""))
		local countO = select(2, flat:gsub("O", ""))
		local winX = win(flat, "X")
		local winO = win(flat, "O")

		assert(countX >= countO)
		if countX > 1 then assert(countO > 0) end
		if winX then assert(countX > countO) end
		if winO then assert(countO >= countX) end

		if winX or winO then return "win" end
		return countX + countO == 9 and "draw" or "ongoing"
	end,
}
